pathway-analysis/03-gene_set_variation_analysis.Rmd

```diff
@@ -150,9 +150,9 @@ gsva_results <- gsva(rnaseq_mat,
                      # Maximum gene set size
                      max.sz = 500,
                      # Compute Gaussian-distributed scores
-                     mx.diff = TRUE,
+                     mx.diff = TRUE)
                      # Use 4 cores
-                     parallel.sz = 4)
+                     # parallel.sz = 4)
```

Comment:
Since this is tuned specifically for multiple people running this the RStudio server, what if we did something like:

```R
ncores_for_server <- 4
if (parallel::detectCores() < ncores_for_server) {
  ncores <- 1
} else {
  ncores <- ncores_for_server
}
```

and then `parallel.sz = 1`? This seems like it could be a useful thing to cover. It might be too _in the weeds_, but I think you're a better judge of that kind of thing than I am!

Reply:
I assume you mean `parallel.sz = ncores`?

I was thinking something just a bit simpler, but same idea:
`ncores <- min(parallel::detectCores(), max_cores)`

Reply:
I can't decided if this is too complex, but I think we could explain it with reasonable facility; this will appear after all toward the end of any given course.

Reply:
Yes, I did mean `parallel.sz = ncores` 😅  your proposed solution works for me!

Reply:
I tried this, and it failed in GHA. It also seems that running in parallel triggers some odd RStudio bug related to (but I guess not exactly the same as?) https://github.com/rstudio/rstudio/issues/7001#issue-627302035 

GSVA seems to run fast enough with one core but it might be worth exploring how to make it work in parallel for other applications. It may be that `parallel.sz` is not sufficient, as it seems like the `BPPARAM` argument may be required (it defaults to `SerialParam`, which might be why increasing the # of cores didn't seem to speed things up much if at all). I think this will require a more substantial investigation, so I am pulling it out to a separate issue.